features/nvstore/source/nvstore.cpp

```diff
@@ -0,0 +1,926 @@
+}
+
+int NVStore::do_set(uint16_t key, uint16_t buf_size, const void *buf, uint16_t flags)
+{
+    int ret = NVSTORE_SUCCESS;
+    uint32_t record_offset, record_size, new_free_space;
+    uint32_t next_offset;
+
+    if (!_init_done) {
+        ret = init();
+        if (ret != NVSTORE_SUCCESS) {
+            return ret;
+        }
+    }
+
+    if (key >= _max_keys) {
+        return NVSTORE_BAD_VALUE;
+    }
+
+    if (!buf) {
+        buf_size = 0;
+    }
+
+    if ((flags & delete_item_flag) && !_offset_by_key[key]) {
+        return NVSTORE_NOT_FOUND;
+    }
+
+    if (_offset_by_key[key] & offs_by_key_set_once_mask) {
+        return NVSTORE_ALREADY_EXISTS;
+    }
+
+    record_size = align_up(sizeof(nvstore_record_header_t) + buf_size, _min_prog_size);
+
+    _mutex->lock();
+
+    new_free_space = core_util_atomic_incr_u32(&_free_space_offset, record_size);
+    record_offset = new_free_space - record_size;
+
+    // If we cross the area limit, we need to invoke GC.
+    if (new_free_space >= _size) {
+        ret = garbage_collection(key, flags, buf_size, buf);
+        _mutex->unlock();
+        return ret;
+    }
+
+    // Now write the record
+    ret = write_record(_active_area, record_offset, key, flags, buf_size, buf, next_offset);
+    if (ret != NVSTORE_SUCCESS) {
+        _mutex->unlock();
+        return ret;
+    }
+
+    // Update _offset_by_key. High bit indicates area.
+    if (flags & delete_item_flag) {
+        _offset_by_key[key] = 0;
+    } else {
+        _offset_by_key[key] = record_offset | (_active_area << offs_by_key_area_bit_pos) |
+                              (((flags & set_once_flag) != 0) << offs_by_key_set_once_bit_pos);
+    }
+
+    _mutex->unlock();
+
+    return NVSTORE_SUCCESS;
+}
+
+int NVStore::set(uint16_t key, uint16_t buf_size, const void *buf)
+{
+    return do_set(key, buf_size, buf, 0);
+}
+
+int NVStore::set_once(uint16_t key, uint16_t buf_size, const void *buf)
+{
+    return do_set(key, buf_size, buf, set_once_flag);
+}
+
+int NVStore::remove(uint16_t key)
+{
+    return do_set(key, 0, NULL, delete_item_flag);
+}
+
+int NVStore::init()
+{
+    area_state_e area_state[NVSTORE_NUM_AREAS];
+    uint32_t free_space_offset_of_area[NVSTORE_NUM_AREAS];
+    uint32_t init_attempts_val;
+    uint32_t next_offset;
+    int os_ret;
+    int ret = NVSTORE_SUCCESS;
+    int valid;
+    uint16_t key;
+    uint16_t flags;
+    uint16_t versions[NVSTORE_NUM_AREAS];
+    uint16_t actual_size;
+
+    if (_init_done) {
+        return NVSTORE_SUCCESS;
+    }
+
+    // This handles the case that init function is called by more than one thread concurrently.
+    // Only the one who gets the value of 1 in _init_attempts_val will proceed, while others will
+    // wait until init is finished.
+    init_attempts_val = core_util_atomic_incr_u32(&_init_attempts, 1);
+    if (init_attempts_val != 1) {
+#ifdef MBED_CONF_RTOS_PRESENT
+        while (!_init_done)
+            rtos::Thread::wait(1);
```

Comment:
just calling wait for 1ms or greater will handle the rtos/non-rtos logic, it shouldn't be needed here.